### PR TITLE
Small feedservice

### PR DESCRIPTION
**Pull request by Kraemii**
**What**:
Small feed service on sensor side. The only command supported is Get VT
Jira: SC-372

**Why**:
For some client it is necessary to get VT information via eulabeia. This small feed service supports the Get VT command which returns all information about a single VT.

**How**:
Start the director and sensor. After openvas is finished loading VTs into redis make a request (e.g. with MQTT Explorer) onto the topic `<context>/vt/cmd/<sensorID>` with the following content:
```json
{
  "created": 0,
  "message_type": "get.feed",
  "message_id": "1",
  "group_id": "1",
  "id": "1.3.6.1.4.1.25623.1.0.120385"
}
```
Instead of the above id any oid stored in redis can be used.

**Checklist**:

- [x] Tests
- [x] [CHANGELOG](https://github.com/greenbone/eulabeia/blob/master/CHANGELOG.md) Entry
- [ ] [DOCUMENTATION](https://github.com/greenbone/eulabeia/tree/main/docs)

### CODE REVIEW

File: feedservice/redis/redis.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gomodule/redigo/redis"
+)
+
+const (
+	NVT_FILENAME_POS           = iota //0
```
Comment (nichtsfrei) [inline feedservice/redis/redis.go:11]: you can assign them manually then you don't have to put a comment to identify the number :) also it's is more explicit than ioata.

**Review by nichtsfrei — CHANGES_REQUESTED** [review]

File: feedservice/feedservice.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	for _, v := range xrefSlice {
+		xref := strings.SplitN(v, ":", 2)
+		if len(xref) != 2 {
+			continue
+		}
+		ret[i] = models.RefType{
+			Type: xref[0],
+			ID:   xref[1],
+		}
+		i++
```
Comment (nichtsfrei) [inline feedservice/feedservice.go:98]: you can get the range with 
```
for i, v :=range xrefSlicee {
```
then you don't need to create i before and i++ it.

Comment (Kraemii) [reply feedservice/feedservice.go:98]: I need the `i` as a counter for `cveSlice`, `bidSlice` and `xrefSlice` as all three slices will be put in a single slice for references.

Comment (nichtsfrei) [reply feedservice/feedservice.go:98]: oh, right my mistake.

File: models/target.go [hunk 3]
```diff
@@ -19,62 +19,75 @@ package models
...
 }
 
 // Severities are stored as Tag. Because tag names can not be repeated, only one severity is supported.
 // The old tag name for severities are cvss_base and cvss_base_vector. The new extended severities tags has priority over the old format.
 type SeverityType struct {
-	SeverityVector  string // CVSS vector. Supported are CVSSv2 and CVSSv3.x
-	SeverityVersion string // CVSS version
-	SeverityDate    string // CVE creation date. Default to VT creation date
-	SeverityOrigin  string // Serverity Origin
+	SeverityVector  string `json:"severity_vector"` // CVSS vector. Supported are CVSSv2 and CVSSv3.x
```
Comment (nichtsfrei) [inline models/target.go:46]: Why do the field names have the prefix Severity* when the struct name is already `SeverityType` and the key in json within VT is also called `severity`?

File: models/target.go [hunk 4]
```diff
@@ -19,62 +19,75 @@ package models
...
+	VTDependencies     []string      `json:"vt_dependencies"`    // List of plugin's filenames which a VT depends on.-
+	Severity           SeverityType  `json:"severety"`           // Script severity. See SeverityType.
+}
+
+type GetVT struct {
+	cmds.EventType
+	messages.Identifier
+}
+
+type SendVT struct {
```
Comment (nichtsfrei) [inline models/target.go:87]: shouldn't that be `GotVT`?

File: feedservice/handler/handler.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+		case "get":
+			vt, err := handler.GetVt(msg.ID)
+			if err != nil {
+				return nil, err
+			}
+
+			return &connection.SendResponse{
+				MSG: models.SendVT{
+					Identifier: messages.Identifier{
+						Message: messages.NewMessage("vts.feed", "", msg.GroupID),
```
Comment (nichtsfrei) [inline feedservice/handler/handler.go:38]: should the message name be `got.vts` or `got.feed` on a get request?

**Review by nichtsfrei — CHANGES_REQUESTED** [review]

File: models/target.go (hunk 3 above)
Comment (Kraemii) [reply models/target.go:46]: I just took it as it was. I change it also for Parameter.

File: feedservice/handler/handler.go (hunk 5 above)
Comment (Kraemii) [reply feedservice/handler/handler.go:38]: I will change the method to `GotVT` and rename the aggregate to vt instead of feed, as it is more accurate.

**Review by nichtsfrei — APPROVED** [review]

**Review by nichtsfrei — DISMISSED** [review]

File: feedservice/feedservice.go [hunk 6]
```diff
@@ -168,11 +168,11 @@ func (f *feed) GetVt(oid string) (models.VT, error) {
 	if timeout != "" {
```
Comment (jjnicola) [inline feedservice/feedservice.go:168]: Not necessary anymore. Timeout is stored in redis and handled as another script_preferences
https://github.com/greenbone/gvm-libs/pull/578

File: feedservice/redis/redis.go [hunk 7]
```diff
@@ -8,21 +8,21 @@ import (
...
+	NVT_EXCLUDED_KEYS_POS      = 3
+	NVT_REQUIRED_UDP_PORTS_POS = 4
+	NVT_REQUIRED_PORTS_POS     = 5
+	NVT_DEPENDENCIES_POS       = 6
+	NVT_TAGS_POS               = 7
+	NVT_CVES_POS               = 8
+	NVT_BIDS_POS               = 9
+	NVT_XREFS_POS              = 10
+	NVT_CATEGORY_POS           = 11
+	NVT_TIMEOUT_POS            = 12
```
Comment (jjnicola) [inline feedservice/redis/redis.go:23]: Not necessary anymore. Timeout is stored in redis and handled as another script_preferences
greenbone/gvm-libs#578

**Review by jjnicola — CHANGES_REQUESTED** [review]
LGTM. Just remove the timeout handling.

**Review by nichtsfrei — APPROVED** [review]

File: feedservice/redis/redis.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	NVT_EXCLUDED_KEYS_POS      = 3
+	NVT_REQUIRED_UDP_PORTS_POS = 4
+	NVT_REQUIRED_PORTS_POS     = 5
+	NVT_DEPENDENCIES_POS       = 6
+	NVT_TAGS_POS               = 7
+	NVT_CVES_POS               = 8
+	NVT_BIDS_POS               = 9
+	NVT_XREFS_POS              = 10
+	NVT_CATEGORY_POS           = 11
+	NVT_FAMILY_POS             = 13
```
Comment (jjnicola) [inline feedservice/redis/redis.go:23]: fix ordinal
```suggestion
	NVT_FAMILY_POS             = 12
	NVT_NAME_POS               =  13
```

Comment (Kraemii) [reply feedservice/redis/redis.go:23]: But the timeout Value is still written in redis and has the index 12. If I adjust the numbers they don't match their corresponding redis entry

Comment (jjnicola) [reply feedservice/redis/redis.go:23]: No, it is not stored anymore in the key. Under the `nvt:<OID>`, there are values from 0-13, and there is no timeout anymore. Timeout is now stored under `oid:<OID>:prefs` , with pref ID 0.

Check
-  https://github.com/greenbone/gvm-libs/blob/05845e35f5e87fe8f513c5d64d3df35c3540b3fe/util/kb.h#L72
- https://github.com/greenbone/gvm-libs/blob/05845e35f5e87fe8f513c5d64d3df35c3540b3fe/util/kb.c#L1561

```
$ redis-cli -s /tmp/redis-openvas.sock 
redis /tmp/redis-openvas.sock> SELECT 1
OK
redis /tmp/redis-openvas.sock[1]> lrange nvt:1.3.6.1.4.1.25623.1.0.902269 0 -1
 1) "2010/secpod_ms10-012-remote.nasl"
 2) ""
 3) "Host/runs_windows"
 4) "SMB/samba"
 5) ""
 6) "445"
 7) "netbios_name_get.nasl, smb_nativelanman.nasl, os_detection.nasl"
 8) "last_modification=1578387992|creation_date=1287755515|cvss_base_vector=AV:N/AC:L/Au:N/C:C/I:C/A:C|impact=Successful exploitation will allow remote attackers to execute arbitrary\n  code or cause a denial of service or bypass the authentication mechanism via brute force technique.|affected=- Microsoft Windows 7\n\n  - Microsoft Windows 2000 Service Pack and prior\n\n  - Microsoft Windows XP Service Pack 3 and prior\n\n  - Microsoft Windows Vista Service Pack 2 and prior\n\n  - Microsoft Windows Server 2003 Service Pack 2 and prior\n\n  - Microsoft Windows Server 2008 Service Pack 2 and prior|insight=- An input validation error exists while processing SMB requests and can\n  be exploited to cause a buffer overflow via a specially crafted SMB packet.\n\n  - An error exists in the SMB implementation while parsing SMB packets during\n  the Negotiate phase causing memory corruption via a specially crafted SMB packet.\n\n  - NULL pointer dereference error exists in SMB while verifying the 'share'\n  and 'servername' fields in SMB packets causing denial of service.\n\n  - A lack of cryptographic entropy when the SMB server generates challenges\n  during SMB NTLM authentication and can be exploited to bypass the authentication mechanism.|solution=The vendor has released updates. Please see the references for more information.|summary=This host is missing a critical security update according to\n  Microsoft Bulletin MS10-012.|qod_type=remote_app|solution_type=VendorFix"
 9) "CVE-2010-0020, CVE-2010-0021, CVE-2010-0022, CVE-2010-0231"
10) ""
11) "URL:http://support.microsoft.com/kb/971468, URL:http://www.vupen.com/english/advisories/2010/0345, URL:https://docs.microsoft.com/en-us/security-updates/securitybulletins/2010/ms10-012"
12) "4"
13) "Windows : Microsoft Bulletins"
14) "Microsoft Windows SMB Server NTLM Multiple Vulnerabilities (971468)"

```

**Review by jjnicola — APPROVED** [review]

**Review by y0urself — APPROVED** [review]

**Pull request merged**